File: spine-libgdx/spine-libgdx/src/com/esotericsoftware/spine/TransformConstraintData.java
/******************************************************************************
 * Spine Runtimes License Agreement
 * Last updated April 5, 2025. Replaces all prior versions.
 *
 * Copyright (c) 2013-2025, Esoteric Software LLC
 *
 * Integration of the Spine Runtimes into software or otherwise creating
 * derivative works of the Spine Runtimes is permitted under the terms and
 * conditions of Section 2 of the Spine Editor License Agreement:
 * http://esotericsoftware.com/spine-editor-license
 *
 * Otherwise, it is permitted to integrate the Spine Runtimes into software
 * or otherwise create derivative works of the Spine Runtimes (collectively,
 * "Products"), provided that each user of the Products must obtain their own
 * Spine Editor license and redistribution of the Products in any form must
 * include this license and copyright notice.
 *
 * THE SPINE RUNTIMES ARE PROVIDED BY ESOTERIC SOFTWARE LLC "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL ESOTERIC SOFTWARE LLC BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES,
 * BUSINESS INTERRUPTION, OR LOSS OF USE, DATA, OR PROFITS) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THE SPINE RUNTIMES, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

package com.esotericsoftware.spine;

import static com.esotericsoftware.spine.utils.SpineUtils.*;

import com.badlogic.gdx.utils.Array;

/** Stores the setup pose for a {@link TransformConstraint}.
 * <p>
 * See <a href="https://esotericsoftware.com/spine-transform-constraints">Transform constraints</a> in the Spine User Guide. */
public class TransformConstraintData extends PosedData<TransformConstraintPose> {
	final Array<BoneData> bones = new Array();
	BoneData source;
	float offsetRotation, offsetX, offsetY, offsetScaleX, offsetScaleY, offsetShearY;
	boolean localSource, localTarget, additive, clamp;
	final Array<FromProperty> properties = new Array();

	public TransformConstraintData (String name) {
		super(name, new TransformConstraintPose());
	}

	/** The bones that will be modified by this transform constraint. */
	public Array<BoneData> getBones () {
		return bones;
	}

	/** The bone whose world transform will be copied to the constrained bones. */
	public BoneData getSource () {
		return source;
	}

	public void setSource (BoneData source) {
		if (source == null) throw new IllegalArgumentException("source cannot be null.");
		this.source = source;
	}

	/** An offset added to the constrained bone rotation. */
	public float getOffsetRotation () {
		return offsetRotation;
	}

	public void setOffsetRotation (float offsetRotation) {
		this.offsetRotation = offsetRotation;
	}

	/** An offset added to the constrained bone X translation. */
	public float getOffsetX () {
		return offsetX;
	}

	public void setOffsetX (float offsetX) {
		this.offsetX = offsetX;
	}

	/** An offset added to the constrained bone Y translation. */
	public float getOffsetY () {
		return offsetY;
	}

	public void setOffsetY (float offsetY) {
		this.offsetY = offsetY;
	}

	/** An offset added to the constrained bone scaleX. */
	public float getOffsetScaleX () {
		return offsetScaleX;
	}

	public void setOffsetScaleX (float offsetScaleX) {
		this.offsetScaleX = offsetScaleX;
	}

	/** An offset added to the constrained bone scaleY. */
	public float getOffsetScaleY () {
		return offsetScaleY;
	}

	public void setOffsetScaleY (float offsetScaleY) {
		this.offsetScaleY = offsetScaleY;
	}

	/** An offset added to the constrained bone shearY. */
	public float getOffsetShearY () {
		return offsetShearY;
	}

	public void setOffsetShearY (float offsetShearY) {
		this.offsetShearY = offsetShearY;
	}

	/** Reads the source bone's local transform instead of its world transform. */
	public boolean getLocalSource () {
		return localSource;
	}

	public void setLocalSource (boolean localSource) {
		this.localSource = localSource;
	}

	/** Sets the constrained bones' local transforms instead of their world transforms. */
	public boolean getLocalTarget () {
		return localTarget;
	}

	public void setLocalTarget (boolean localTarget) {
		this.localTarget = localTarget;
	}

	/** Adds the source bone transform to the constrained bones instead of setting it absolutely. */
	public boolean getAdditive () {
		return additive;
	}

	public void setAdditive (boolean additive) {
		this.additive = additive;
	}

	/** Prevents constrained bones from exceeding the ranged defined by {@link ToProperty#offset} and {@link ToProperty#max}. */
	public boolean getClamp () {
		return clamp;
	}

	public void setClamp (boolean clamp) {
		this.clamp = clamp;
	}

	/** The mapping of transform properties to other transform properties. */
	public Array<FromProperty> getProperties () {
		return properties;
	}

	/** Source property for a {@link TransformConstraint}. */
	static abstract public class FromProperty {
		/** The value of this property that corresponds to {@link ToProperty#offset}. */
		public float offset;

		/** Constrained properties. */
		public final Array<ToProperty> to = new Array();

		/** Reads this property from the specified bone. */
		abstract public float value (TransformConstraintData data, BonePose source, boolean local);
	}

	/** Constrained property for a {@link TransformConstraint}. */
	static abstract public class ToProperty {
		/** The value of this property that corresponds to {@link FromProperty#offset}. */
		public float offset;

		/** The maximum value of this property when {@link TransformConstraintData#clamp clamped}. */
		public float max;

		/** The scale of the {@link FromProperty} value in relation to this property. */
		public float scale;

		/** Reads the mix for this property from the specified pose. */
		abstract public float mix (TransformConstraintPose pose);

		/** Applies the value to this property. */
		abstract public void apply (TransformConstraintPose pose, BonePose bone, float value, boolean local, boolean additive);
	}

	static public class FromRotate extends FromProperty {
		public float value (TransformConstraintData data, BonePose source, boolean local) {
			if (local) return source.rotation + data.offsetRotation;
			float value = atan2(source.c, source.a) * radDeg
				+ (source.a * source.d - source.b * source.c > 0 ? data.offsetRotation : -data.offsetRotation);
			if (value < 0) value += 360;
			return value;
		}
	}

	static public class ToRotate extends ToProperty {
		public float mix (TransformConstraintPose pose) {
			return pose.mixRotate;
		}

		public void apply (TransformConstraintPose pose, BonePose bone, float value, boolean local, boolean additive) {
			if (local) {
				if (!additive) value -= bone.rotation;
				bone.rotation += value * pose.mixRotate;
			} else {
				float a = bone.a, b = bone.b, c = bone.c, d = bone.d;
				value *= degRad;
				if (!additive) value -= atan2(c, a);
				if (value > PI)
					value -= PI2;
				else if (value < -PI) //
					value += PI2;
				value *= pose.mixRotate;
				float cos = cos(value), sin = sin(value);
				bone.a = cos * a - sin * c;
				bone.b = cos * b - sin * d;
				bone.c = sin * a + cos * c;
				bone.d = sin * b + cos * d;
			}
		}
	}

	static public class FromX extends FromProperty {
		public float value (TransformConstraintData data, BonePose source, boolean local) {
			return local ? source.x + data.offsetX : data.offsetX * source.a + data.offsetY * source.b + source.worldX;
		}
	}

	static public class ToX extends ToProperty {
		public float mix (TransformConstraintPose pose) {
			return pose.mixX;
		}

		public void apply (TransformConstraintPose pose, BonePose bone, float value, boolean local, boolean additive) {
			if (local) {
				if (!additive) value -= bone.x;
				bone.x += value * pose.mixX;
			} else {
				if (!additive) value -= bone.worldX;
				bone.worldX += value * pose.mixX;
			}
		}
	}

	static public class FromY extends FromProperty {
		public float value (TransformConstraintData data, BonePose source, boolean local) {
			return local ? source.y + data.offsetY : data.offsetX * source.c + data.offsetY * source.d + source.worldY;
		}
	}

	static public class ToY extends ToProperty {
		public float mix (TransformConstraintPose pose) {
			return pose.mixY;
		}

		public void apply (TransformConstraintPose pose, BonePose bone, float value, boolean local, boolean additive) {
			if (local) {
				if (!additive) value -= bone.y;
				bone.y += value * pose.mixY;
			} else {
				if (!additive) value -= bone.worldY;
				bone.worldY += value * pose.mixY;
			}
		}
	}

	static public class FromScaleX extends FromProperty {
		public float value (TransformConstraintData data, BonePose source, boolean local) {
			return (local ? source.scaleX : (float)Math.sqrt(source.a * source.a + source.c * source.c)) + data.offsetScaleX;
		}
	}

	static public class ToScaleX extends ToProperty {
		public float mix (TransformConstraintPose pose) {
			return pose.mixScaleX;
		}

		public void apply (TransformConstraintPose pose, BonePose bone, float value, boolean local, boolean additive) {
			if (local) {
				if (additive)
					bone.scaleX *= 1 + ((value - 1) * pose.mixScaleX);
				else if (bone.scaleX != 0) //
					bone.scaleX = 1 + (value / bone.scaleX - 1) * pose.mixScaleX;
			} else {
				float s;
				if (additive)
					s = 1 + (value - 1) * pose.mixScaleX;
				else {
					s = (float)Math.sqrt(bone.a * bone.a + bone.c * bone.c);
					if (s != 0) s = 1 + (value / s - 1) * pose.mixScaleX;
				}
				bone.a *= s;
				bone.c *= s;
			}
		}
	}

	static public class FromScaleY extends FromProperty {
		public float value (TransformConstraintData data, BonePose source, boolean local) {
			return (local ? source.scaleY : (float)Math.sqrt(source.b * source.b + source.d * source.d)) + data.offsetScaleY;
		}
	}

	static public class ToScaleY extends ToProperty {
		public float mix (TransformConstraintPose pose) {
			return pose.mixScaleY;
		}

		public void apply (TransformConstraintPose pose, BonePose bone, float value, boolean local, boolean additive) {
			if (local) {
				if (additive)
					bone.scaleY *= 1 + ((value - 1) * pose.mixScaleY);
				else if (bone.scaleY != 0) //
					bone.scaleY = 1 + (value / bone.scaleY - 1) * pose.mixScaleY;
			} else {
				float s;
				if (additive)
					s = 1 + (value - 1) * pose.mixScaleY;
				else {
					s = (float)Math.sqrt(bone.b * bone.b + bone.d * bone.d);
					if (s != 0) s = 1 + (value / s - 1) * pose.mixScaleY;
				}
				bone.b *= s;
				bone.d *= s;
			}
		}
	}

	static public class FromShearY extends FromProperty {
		public float value (TransformConstraintData data, BonePose source, boolean local) {
			return (local ? source.shearY : (atan2(source.d, source.b) - atan2(source.c, source.a)) * radDeg - 90)
				+ data.offsetShearY;
		}
	}

	static public class ToShearY extends ToProperty {
		public float mix (TransformConstraintPose pose) {
			return pose.mixShearY;
		}

		public void apply (TransformConstraintPose pose, BonePose bone, float value, boolean local, boolean additive) {
			if (local) {
				if (!additive) value -= bone.shearY;
				bone.shearY += value * pose.mixShearY;
			} else {
				float b = bone.b, d = bone.d, by = atan2(d, b);
				value = (value + 90) * degRad;
				if (additive)
					value -= PI / 2;
				else {
					value -= by - atan2(bone.c, bone.a);
					if (value > PI)
						value -= PI2;
					else if (value < -PI) //
						value += PI2;
				}
				value = by + value * pose.mixShearY;
				float s = (float)Math.sqrt(b * b + d * d);
				bone.b = cos(value) * s;
				bone.d = sin(value) * s;
			}
		}
	}
}

File: spine-ts/spine-core/src/TransformConstraintData.ts
/******************************************************************************
 * Spine Runtimes License Agreement
 * Last updated April 5, 2025. Replaces all prior versions.
 *
 * Copyright (c) 2013-2025, Esoteric Software LLC
 *
 * Integration of the Spine Runtimes into software or otherwise creating
 * derivative works of the Spine Runtimes is permitted under the terms and
 * conditions of Section 2 of the Spine Editor License Agreement:
 * http://esotericsoftware.com/spine-editor-license
 *
 * Otherwise, it is permitted to integrate the Spine Runtimes into software
 * or otherwise create derivative works of the Spine Runtimes (collectively,
 * "Products"), provided that each user of the Products must obtain their own
 * Spine Editor license and redistribution of the Products in any form must
 * include this license and copyright notice.
 *
 * THE SPINE RUNTIMES ARE PROVIDED BY ESOTERIC SOFTWARE LLC "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL ESOTERIC SOFTWARE LLC BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES,
 * BUSINESS INTERRUPTION, OR LOSS OF USE, DATA, OR PROFITS) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THE SPINE RUNTIMES, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

import { ConstraintData } from "./ConstraintData.js";
import { BoneData } from "./BoneData.js";
import { TransformConstraint } from "./TransformConstraint.js";
import { MathUtils } from "./Utils.js";
import { Skeleton } from "./Skeleton.js";
import { TransformConstraintPose } from "./TransformConstraintPose.js";
import { BonePose } from "./BonePose.js";

/** Stores the setup pose for a {@link TransformConstraint}.
 *
 * See [Transform constraints](http://esotericsoftware.com/spine-transform-constraints) in the Spine User Guide. */
export class TransformConstraintData extends ConstraintData<TransformConstraint, TransformConstraintPose> {
	public static readonly ROTATION = 0;
	public static readonly X = 1;
	public static readonly Y = 2;
	public static readonly SCALEX = 3;
	public static readonly SCALEY = 4;
	public static readonly SHEARY = 5;

	/** The bones that will be modified by this transform constraint. */
	bones = new Array<BoneData>();

	/** The bone whose world transform will be copied to the constrained bones. */
	public set source (source: BoneData) { this._source = source; }
	public get source () {
		if (!this._source) throw new Error("BoneData not set.")
		else return this._source;
	}
	private _source: BoneData | null = null;

	offsets = [0, 0, 0, 0, 0, 0];

	/** An offset added to the constrained bone X translation. */
	offsetX = 0;

	/** An offset added to the constrained bone Y translation. */
	offsetY = 0;

	/** Reads the source bone's local transform instead of its world transform. */
	localSource = false;

	/** Sets the constrained bones' local transforms instead of their world transforms. */
	localTarget = false;

	/** Adds the source bone transform to the constrained bones instead of setting it absolutely. */
	additive = false;

	/** Prevents constrained bones from exceeding the ranged defined by {@link ToProperty.offset} and {@link ToProperty.max}. */
	clamp = false;

	/** The mapping of transform properties to other transform properties. */
	readonly properties: Array<FromProperty> = [];

	constructor (name: string) {
		super(name, new TransformConstraintPose());
	}

	public create (skeleton: Skeleton) {
		return new TransformConstraint(this, skeleton);
	}

	/** An offset added to the constrained bone rotation. */
	getOffsetRotation () {
		return this.offsets[TransformConstraintData.ROTATION];
	}

	setOffsetRotation (offsetRotation: number) {
		this.offsets[TransformConstraintData.ROTATION] = offsetRotation;
	}

	/** An offset added to the constrained bone X translation. */
	getOffsetX () {
		return this.offsets[TransformConstraintData.X];
	}

	setOffsetX (offsetX: number) {
		this.offsets[TransformConstraintData.X] = offsetX;
	}

	/** An offset added to the constrained bone Y translation. */
	getOffsetY () {
		return this.offsets[TransformConstraintData.Y];
	}

	setOffsetY (offsetY: number) {
		this.offsets[TransformConstraintData.Y] = offsetY;
	}

	/** An offset added to the constrained bone scaleX. */
	getOffsetScaleX () {
		return this.offsets[TransformConstraintData.SCALEX];
	}

	setOffsetScaleX (offsetScaleX: number) {
		this.offsets[TransformConstraintData.SCALEX] = offsetScaleX;
	}

	/** An offset added to the constrained bone scaleY. */
	getOffsetScaleY () {
		return this.offsets[TransformConstraintData.SCALEY];
	}

	setOffsetScaleY (offsetScaleY: number) {
		this.offsets[TransformConstraintData.SCALEY] = offsetScaleY;
	}

	/** An offset added to the constrained bone shearY. */
	getOffsetShearY () {
		return this.offsets[TransformConstraintData.SHEARY];
	}

	setOffsetShearY (offsetShearY: number) {
		this.offsets[TransformConstraintData.SHEARY] = offsetShearY;
	}

}

/** Source property for a {@link TransformConstraint}. */
export abstract class FromProperty {
	/** The value of this property that corresponds to {@link ToProperty#offset}. */
	offset = 0;

	/** Constrained properties. */
	readonly to: Array<ToProperty> = [];

	/** Reads this property from the specified bone. */
	abstract value (skeleton: Skeleton, source: BonePose, local: boolean, offsets: Array<number>): number;
}

/** Constrained property for a {@link TransformConstraint}. */
export abstract class ToProperty {
	/** The value of this property that corresponds to {@link FromProperty#offset}. */
	offset = 0;

	/** The maximum value of this property when {@link TransformConstraintData#clamp clamped}. */
	max = 0;

	/** The scale of the {@link FromProperty} value in relation to this property. */
	scale = 0;

	/** Reads the mix for this property from the specified constraint. */
	abstract mix (pose: TransformConstraintPose): number;

	/** Applies the value to this property. */
	abstract apply (skeleton: Skeleton, pose: TransformConstraintPose, bone: BonePose, value: number, local: boolean, additive: boolean): void;
}

export class FromRotate extends FromProperty {
	value (skeleton: Skeleton, source: BonePose, local: boolean, offsets: Array<number>): number {
		if (local) return source.rotation + offsets[TransformConstraintData.ROTATION];
		let value = Math.atan2(source.c / skeleton.scaleY, source.a / skeleton.scaleX) * MathUtils.radDeg
			+ (source.a * source.d - source.b * source.c > 0 ? offsets[TransformConstraintData.ROTATION] : -offsets[TransformConstraintData.ROTATION]);
		if (value < 0) value += 360;
		return value;
	}
}

export class ToRotate extends ToProperty {
	mix (pose: TransformConstraintPose): number {
		return pose.mixRotate;
	}

	apply (skeleton: Skeleton, pose: TransformConstraintPose, bone: BonePose, value: number, local: boolean, additive: boolean): void {
		if (local) {
			if (!additive) value -= bone.rotation;
			bone.rotation += value * pose.mixRotate;
		} else {
			const a = bone.a, b = bone.b, c = bone.c, d = bone.d;
			value *= MathUtils.degRad;
			if (!additive) value -= Math.atan2(c, a);
			if (value > MathUtils.PI)
				value -= MathUtils.PI2;
			else if (value < -MathUtils.PI) //
				value += MathUtils.PI2;
			value *= pose.mixRotate;
			const cos = Math.cos(value), sin = Math.sin(value);
			bone.a = cos * a - sin * c;
			bone.b = cos * b - sin * d;
			bone.c = sin * a + cos * c;
			bone.d = sin * b + cos * d;
		}
	}
}

export class FromX extends FromProperty {
	value (skeleton: Skeleton, source: BonePose, local: boolean, offsets: Array<number>): number {
		return local
			? source.x + offsets[TransformConstraintData.X]
			: (offsets[TransformConstraintData.X] * source.a + offsets[TransformConstraintData.Y] * source.b + source.worldX) / skeleton.scaleX;
	}
}

export class ToX extends ToProperty {
	mix (pose: TransformConstraintPose): number {
		return pose.mixX;
	}

	apply (skeleton: Skeleton, pose: TransformConstraintPose, bone: BonePose, value: number, local: boolean, additive: boolean): void {
		if (local) {
			if (!additive) value -= bone.x;
			bone.x += value * pose.mixX;
		} else {
			if (!additive) value -= bone.worldX;
			bone.worldX += value * pose.mixX;
		}
	}
}

export class FromY extends FromProperty {
	value (skeleton: Skeleton, source: BonePose, local: boolean, offsets: Array<number>): number {
		return local
			? source.y + offsets[TransformConstraintData.Y]
			: (offsets[TransformConstraintData.X] * source.c + offsets[TransformConstraintData.Y] * source.d + source.worldY) / skeleton.scaleY;
	}
}

export class ToY extends ToProperty {
	mix (pose: TransformConstraintPose): number {
		return pose.mixY;
	}

	apply (skeleton: Skeleton, pose: TransformConstraintPose, bone: BonePose, value: number, local: boolean, additive: boolean): void {
		if (local) {
			if (!additive) value -= bone.y;
			bone.y += value * pose.mixY;
		} else {
			if (!additive) value -= bone.worldY;
			bone.worldY += value * pose.mixY;
		}
	}
}

export class FromScaleX extends FromProperty {
	value (skeleton: Skeleton, source: BonePose, local: boolean, offsets: Array<number>): number {
		if (local) return source.scaleX + offsets[TransformConstraintData.SCALEX];
		const a = source.a / skeleton.scaleX, c = source.c / skeleton.scaleY;
		return Math.sqrt(a * a + c * c) + offsets[TransformConstraintData.SCALEX];
	}
}

export class ToScaleX extends ToProperty {
	mix (pose: TransformConstraintPose): number {
		return pose.mixScaleX;
	}

	apply (skeleton: Skeleton, pose: TransformConstraintPose, bone: BonePose, value: number, local: boolean, additive: boolean): void {
		if (local) {
			if (additive)
				bone.scaleX *= 1 + ((value - 1) * pose.mixScaleX);
			else if (bone.scaleX != 0) //
				bone.scaleX = 1 + (value / bone.scaleX - 1) * pose.mixScaleX;
		} else {
			let s: number;
			if (additive)
				s = 1 + (value - 1) * pose.mixScaleX;
			else {
				s = Math.sqrt(bone.a * bone.a + bone.c * bone.c) / skeleton.scaleX;
				if (s != 0) s = 1 + (value / s - 1) * pose.mixScaleX;
			}
			bone.a *= s;
			bone.c *= s;
		}
	}
}

export class FromScaleY extends FromProperty {
	value (skeleton: Skeleton, source: BonePose, local: boolean, offsets: Array<number>): number {
		if (local) return source.scaleY + offsets[TransformConstraintData.SCALEY];
		const b = source.b / skeleton.scaleX, d = source.d / skeleton.scaleY;
		return Math.sqrt(b * b + d * d) + offsets[TransformConstraintData.SCALEY];
	}
}

export class ToScaleY extends ToProperty {
	mix (pose: TransformConstraintPose): number {
		return pose.mixScaleY;
	}

	apply (skeleton: Skeleton, pose: TransformConstraintPose, bone: BonePose, value: number, local: boolean, additive: boolean): void {
		if (local) {
			if (additive)
				bone.scaleY *= 1 + ((value - 1) * pose.mixScaleY);
			else if (bone.scaleY != 0) //
				bone.scaleY = 1 + (value / bone.scaleY - 1) * pose.mixScaleY;
		} else {
			let s: number;
			if (additive)
				s = 1 + (value - 1) * pose.mixScaleY;
			else {
				s = Math.sqrt(bone.b * bone.b + bone.d * bone.d) / skeleton.scaleY;
				if (s != 0) s = 1 + (value / s - 1) * pose.mixScaleY;
			}
			bone.b *= s;
			bone.d *= s;
		}
	}
}

export class FromShearY extends FromProperty {
	value (skeleton: Skeleton, source: BonePose, local: boolean, offsets: Array<number>): number {
		if (local) return source.shearY + offsets[TransformConstraintData.SHEARY];
		const sx = 1 / skeleton.scaleX, sy = 1 / skeleton.scaleY;
		return (Math.atan2(source.d * sy, source.b * sx) - Math.atan2(source.c * sy, source.a * sx))
			* MathUtils.radDeg - 90 + offsets[TransformConstraintData.SHEARY];
	}
}

export class ToShearY extends ToProperty {
	mix (pose: TransformConstraintPose): number {
		return pose.mixShearY;
	}

	apply (skeleton: Skeleton, pose: TransformConstraintPose, bone: BonePose, value: number, local: boolean, additive: boolean): void {
		if (local) {
			if (!additive) value -= bone.shearY;
			bone.shearY += value * pose.mixShearY;
		} else {
			const b = bone.b, d = bone.d, by = Math.atan2(d, b);
			value = (value + 90) * MathUtils.degRad;
			if (additive)
				value -= MathUtils.PI / 2;
			else {
				value -= by - Math.atan2(bone.c, bone.a);
				if (value > MathUtils.PI)
					value -= MathUtils.PI2;
				else if (value < -MathUtils.PI)
					value += MathUtils.PI2;
			}
			value = by + value * pose.mixShearY;
			const s = Math.sqrt(b * b + d * d);
			bone.b = Math.cos(value) * s;
			bone.d = Math.sin(value) * s;
		}
	}
}
